File: Java/382. Linked List Random Node.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    private ListNode head;
    
    public Solution(ListNode head) {
        this.head = head;
    }
    
    public int getRandom() {
        int scope = 1, chosen = 0;
        ListNode cur = this.head;
        while (cur != null) {
            if (Math.random() < 1.0 / scope) {
                chosen = cur.val;
            }
            cur = cur.next;
            scope++;
        }
        return chosen;
    }
}

/**
 * Your Solution object will be instantiated and called as such:
 * Solution obj = new Solution(head);
 * int param_1 = obj.getRandom();
 */

File: Python3/382. Linked List Random Node.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:

    def __init__(self, head: Optional[ListNode]):
        self.head = head

    def getRandom(self) -> int:
        cur = self.head
        chosen = 0
        scope = 1

        while cur:
            if random.random() < 1/scope:
                chosen = cur.val
            cur = cur.next
            scope += 1
        
        return chosen


# Your Solution object will be instantiated and called as such:
# obj = Solution(head)
# param_1 = obj.getRandom()